swarm/animator.py

```diff
@@ -45,3 +59,306 @@ def animate(i):
+        pass
+
+
+def _validate_ax_kwargs(**kwargs):
+    """
+    This is an attempt to validate the kwargs on definition rather than at runtime for quicker
+    feedback to the developer.
+    This isn't needed by end-user, but could be used ahead of time if they like.
+
+    Raises:
+        ValueError: in the case of invalid arguments
+    """
+    cls = plt.Axes
+    missing = []
+    for k in kwargs:
+        if not hasattr(cls, k):
+            missing.append(k)
+    if missing:
+        raise ValueError("plt.Axes does not have {}".format(",".join(missing)))
+
+
+def _kwargs_hook(**kwargs):
+    """
+    Our animation library uses hooks to allow once of configuration of an axes. These
+    include things like plotting the true y value or setting sensible x/y limits for the plot
+
+    This can be time consuming to do, so this provides an easy way to do so, allowing
+    one to call the various functions of ``plt.Axes`` as below. This is expected to be fit
+    into the existing machinery of the SwarmPlot sub/duck classes so you won't need to make
+    this call directly
+
+    Examples:
+        >>> _kwargs_hook(set_title="Goofy Experiment", set_ylabel="loss")
+        # will be used mostly with constructors like
+        >>> LineSwarm.standard(x, y, data, set_title="Goofy Experiment", set_ylabel="loss")
+
+    Raises:
+        ValueError: will attempt to validate the kwargs ahead of time, but also on execution
+        in case the function arg is incorrect
+
+    """
+    # get an early error if this is a problem
+    # rather than at animate time
+    _validate_ax_kwargs(**kwargs)
+
+    def inner_hook(ax):
+        for k, v in kwargs.items():
+            try:
+                func = getattr(ax, k)
+            except AttributeError:
+                raise ValueError(f"Used a method {k}")
+            else:
+                try:
+                    func(v)
+                except Exception as e:
+                    # we might have an issue that we're hitting a property
+                    # or have type mismatches
+                    raise ValueError("Unexpected failure") from e
+
+    return inner_hook
+
+
+def _apply_hook(ax: plt.Axes, hook: Union[Iterable[Callable], Callable, None]):
+    """Applies a list of hooks to the axes object"""
+    if not hook:
+        return
+    try:
+        for h in hook:
+            h(ax)
+    except TypeError:
+        # can't iterate, must be single callabe
+        hook(ax)
+
+
+@attr.s
+class LineSwarm(SwarmPlot):
+    """
+    This is the standard class
+    """
+
+    x = attr.ib(type=np.ndarray)  # shape of (N,)
+    data = attr.ib(type=np.ndarray)  # shape of (num_lines, timestep, N)
+    hook = attr.ib(type=List[Callable], factory=list)
+
+    @classmethod
+    def standard(cls, xd, yd, data, **kwargs):
+        """Plot against true yd"""
+        hook = [lambda ax: ax.plot(xd, yd, "."), _kwargs_hook(**kwargs)]
+        return cls(xd, data, hook)
+
+    @classmethod
```

Comment:
My approach to class design is to separate the storage of state from the creation. I.e. I want my `__init__` to do nothing other than store the attributes I care about. 

See the two different approaches
```python
class Layer:
    def __init__(self, inwidth, outwidth, init_method='xavier'):
        if init_method == 'xavier':
            self.W = np.random.random(outwidth, inwidth)
            self.bias = np.random(outwidth)
        else:
            self.W = np.zeros(outwidth, inwidth)
            self.bias = np.zeros(outwidth)


class Layer:
    def __init__(self, weight, bias):
        self.W = weight
        self.bias = bias

    @classmethod
    def xavier(cls, inwidth, outwidth):
        w = np.random.random(outwidth, inwidth)
        bs = np.random(outwidth)
        return cls(w, b)

    @classmethod
    def zero(cls, inwidth, outwidth):
        w = np.zeros(outwidth, inwidth)
        bs = np.zeros(outwidth)
        return cls(w, b)
```

In both approaches the state of the class is the same, but in the first approach has a lot of logic in the class initialisation and repeats code creating `self.W` and `self.bias` each time. The second approach realises this and separates the state of the class from the instantiation. This way you can create the class from a known state with ease (this is useful when recovering state and especially testing) while you still have clear code that describes how a class is created. For example if you had a new_initialiser that took an arg called lambda, you could just have `def new_initialiser(cls, inwidth, outwidth, lambda):` and your `__init__` doesn't need to change to accept an argument that is useless for most initial users.

Attrs takes this a step forward by removing the `__init__` entirely and suggesting classmethods for all creation, which makes python a lot more readable. This was considered such a good idea, that python implemented a subset of attrs as dataclasses: https://realpython.com/python-data-classes/ 

At the end of the day, this is definitely leading into style and the truth is that while I think this approach is better and more readable, not doing it isn't that much of a sin.

FInally `@property` is syntactic sugar. I can do `hist.num_frames` instead of `hist.num_frames()` which is a little nicer, but again not necessary.

Reply:
Thanks, this is cool, I'm still trying to understand how these `@classmethods` work though... I can't see exactly how instantiating `Layer(input_1, input_2)` gives you the same thing in both cases.  Is that what you were proposing? in the second example, how doe the class methods even get called? How is it known that `xavier` is preferred? (Just because it's first?)

Reply:
Calling eg1 `Layer_1` and eg2 `Layer_2`. `Layer_1(inwidth, outwidth) == Layer_2.xavier(inwidth, outwidth)`. That's also how you call classmethods/staticmethods.

As to how to know what to call, you can make it implicit by having a `@classmethod` which is named `def default_init(cls, ...)` that is well documented, or even a function called `make_layer(inwidth, outwidth)` which returns `Layer_2.xavier(inwidth, outwidth)`.